File: mobile/ios/AstraMobile/APIClient.swift
import Foundation
import Security

final class APIClient {
    private let defaults = UserDefaults.standard
    private let baseURLKey = "astra_api_base_url"
    private let tokenKey = "astra_access_token"
    private let defaultBaseURL = "http://204.168.255.69:8787/api/v1"
    private let legacyLocalBaseURLs = [
        "http://127.0.0.1:8787/api/v1",
        "http://10.0.2.2:8787/api/v1"
    ]

    var baseURL: String {
        get {
            let stored = defaults.string(forKey: baseURLKey)
            return stored == nil || legacyLocalBaseURLs.contains(stored ?? "") ? defaultBaseURL : stored!
        }
        set { defaults.set(newValue.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/")), forKey: baseURLKey) }
    }

    var token: String? {
        get {
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrAccount as String: tokenKey,
                kSecReturnData as String: true,
                kSecMatchLimit as String: kSecMatchLimitOne
            ]
            var result: AnyObject?
            guard SecItemCopyMatching(query as CFDictionary, &result) == errSecSuccess,
                  let data = result as? Data else { return nil }
            return String(data: data, encoding: .utf8)
        }
        set {
            let query: [String: Any] = [
                kSecClass as String: kSecClassGenericPassword,
                kSecAttrAccount as String: tokenKey
            ]
            SecItemDelete(query as CFDictionary)
            if let newValue, let data = newValue.data(using: .utf8) {
                var item = query
                item[kSecValueData as String] = data
                SecItemAdd(item as CFDictionary, nil)
            }
        }
    }

    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return decoder
    }()

    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        return encoder
    }()

    func request<T: Decodable>(_ path: String, method: String = "GET", body: Encodable? = nil) async throws -> T {
        guard let url = URL(string: "\(baseURL)/\(path)") else { throw APIError.invalidURL }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 30
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try encoder.encode(AnyEncodable(body))
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidResponse }
        if !(200..<300).contains(http.statusCode) {
            let payload = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            throw APIError.server(payload?["error"] as? String ?? "Ошибка запроса (\(http.statusCode))")
        }
        do { return try decoder.decode(T.self, from: data) }
        catch { throw APIError.decoding(error.localizedDescription) }
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        try await request("auth/login", method: "POST", body: AuthBody(email: email, password: password))
    }

    func register(email: String, password: String) async throws -> AuthResponse {
        try await request("auth/register", method: "POST", body: AuthBody(email: email, password: password))
    }

    func me() async throws -> AuthUser { try await request("auth/me") }
    func dashboard() async throws -> Dashboard { try await request("dashboard") }
    func products() async throws -> [Product] { try await request("products") }
    func recipes() async throws -> [Recipe] { try await request("recipes") }
    func recipe(id: Int) async throws -> RecipeDetail { try await request("recipes/\(id)") }
    func diary() async throws -> [DiaryEntry] { try await request("diary") }
    func progress() async throws -> [ProgressEntry] { try await request("progress") }
    func workouts() async throws -> [WorkoutEntry] { try await request("workouts") }
    func workoutPlans() async throws -> [WorkoutPlan] { try await request("workout-plans") }
    func exercises() async throws -> [Exercise] { try await request("exercises") }
    func workoutEquipment() async throws -> [WorkoutEquipment] { try await request("workout-equipment") }
    func workoutComplexes() async throws -> [WorkoutComplex] { try await request("workout-complexes") }
    func categories(kind: String) async throws -> [ContentCategory] { try await request("categories?kind=\(kind.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? kind)") }
    func articleSections() async throws -> [ArticleSection] { try await request("article-sections") }
    func articles() async throws -> [Article] { try await request("articles") }
    func myTrainer() async throws -> TrainerInfoResponse { try await request("clients/me/trainer") }
    func myTrainerChat() async throws -> TrainerChatResponse { try await request("clients/me/chat") }
    func sendMyTrainerChat(message: String) async throws -> TrainerChatMessage { try await request("clients/me/chat", method: "POST", body: TrainerChatPayload(message: message)) }
    func clients() async throws -> [ClientSummary] { try await request("clients") }
    func client(id: Int) async throws -> ClientDetail { try await request("clients/\(id)") }
    func clientChat(id: Int) async throws -> [TrainerChatMessage] { try await request("clients/\(id)/chat") }
    func sendClientChat(id: Int, message: String) async throws -> TrainerChatMessage { try await request("clients/\(id)/chat", method: "POST", body: TrainerChatPayload(message: message)) }
    func createProduct(_ payload: JSONPayload) async throws -> Product { try await request("products", method: "POST", body: payload) }
    func updateProduct(id: Int, payload: JSONPayload) async throws -> Product { try await request("products/\(id)", method: "PUT", body: payload) }
    func deleteProduct(id: Int) async throws -> DeleteResponse { try await request("products/\(id)", method: "DELETE") }
    func createRecipe(_ payload: JSONPayload) async throws -> Recipe { try await request("recipes", method: "POST", body: payload) }
    func updateRecipe(id: Int, payload: JSONPayload) async throws -> Recipe { try await request("recipes/\(id)", method: "PUT", body: payload) }
    func deleteRecipe(id: Int) async throws -> DeleteResponse { try await request("recipes/\(id)", method: "DELETE") }
    func requestRecipeSubmission(id: Int) async throws -> Recipe { try await request("recipes/\(id)/submission-request", method: "POST") }
    func cancelRecipeSubmission(id: Int) async throws -> Recipe { try await request("recipes/\(id)/submission-request", method: "DELETE") }
    func moderateRecipe(id: Int, action: String, note: String?) async throws -> Recipe { try await request("recipes/\(id)/moderation", method: "POST", body: JSONPayload(values: ["action": AnyEncodable(action), "note": AnyEncodable(note)])) }
    func createExercise(_ payload: JSONPayload) async throws -> Exercise { try await request("exercises", method: "POST", body: payload) }
    func updateExercise(id: Int, payload: JSONPayload) async throws -> Exercise { try await request("exercises/\(id)", method: "PUT", body: payload) }
    func deleteExercise(id: Int) async throws -> DeleteResponse { try await request("exercises/\(id)", method: "DELETE") }
    func createEquipment(_ payload: JSONPayload) async throws -> WorkoutEquipment { try await request("workout-equipment", method: "POST", body: payload) }
    func updateEquipment(id: Int, payload: JSONPayload) async throws -> WorkoutEquipment { try await request("workout-equipment/\(id)", method: "PUT", body: payload) }
    func deleteEquipment(id: Int) async throws -> DeleteResponse { try await request("workout-equipment/\(id)", method: "DELETE") }
    func createComplex(_ payload: JSONPayload) async throws -> WorkoutComplex { try await request("workout-complexes", method: "POST", body: payload) }
    func updateComplex(id: Int, payload: JSONPayload) async throws -> WorkoutComplex { try await request("workout-complexes/\(id)", method: "PUT", body: payload) }
    func deleteComplex(id: Int) async throws -> DeleteResponse { try await request("workout-complexes/\(id)", method: "DELETE") }
    func updateWorkout(id: Int, payload: JSONPayload) async throws -> WorkoutEntry { try await request("workouts/\(id)", method: "PUT", body: payload) }
    func deleteWorkout(id: Int) async throws -> DeleteResponse { try await request("workouts/\(id)", method: "DELETE") }
    func createWorkoutPlan(_ payload: JSONPayload) async throws -> WorkoutPlan { try await request("workout-plans", method: "POST", body: payload) }
    func updateWorkoutPlan(id: Int, payload: JSONPayload) async throws -> WorkoutPlan { try await request("workout-plans/\(id)", method: "PUT", body: payload) }
    func cancelWorkoutPlan(id: Int) async throws -> WorkoutPlan { try await request("workout-plans/\(id)/cancel", method: "POST") }
    func deleteWorkoutPlan(id: Int) async throws -> DeleteResponse { try await request("workout-plans/\(id)", method: "DELETE") }
    func createArticle(_ payload: JSONPayload) async throws -> Article { try await request("articles", method: "POST", body: payload) }
    func updateArticle(id: Int, payload: JSONPayload) async throws -> Article { try await request("articles/\(id)", method: "PUT", body: payload) }
    func updateArticleFlags(id: Int, payload: JSONPayload) async throws -> Article { try await request("articles/\(id)/flags", method: "PATCH", body: payload) }
    func deleteArticle(id: Int) async throws -> DeleteResponse { try await request("articles/\(id)", method: "DELETE") }
    func shareToTrainer(itemType: String, itemId: Int) async throws -> SharedItemResult { try await request("clients/me/shares", method: "POST", body: JSONPayload(values: ["item_type": AnyEncodable(itemType), "item_id": AnyEncodable(itemId)])) }
    func shareToClient(clientId: Int, itemType: String, itemId: Int) async throws -> SharedItemResult { try await request("clients/shares", method: "POST", body: JSONPayload(values: ["client_id": AnyEncodable(clientId), "item_type": AnyEncodable(itemType), "item_id": AnyEncodable(itemId)])) }
    func createDiary(_ payload: DiaryPayload) async throws -> [DiaryEntry] { try await request("diary", method: "POST", body: payload) }
    func createDiary(_ payload: JSONPayload) async throws -> [DiaryEntry] { try await request("diary", method: "POST", body: payload) }
    func updateDiary(id: Int, payload: JSONPayload) async throws -> DiaryEntry { try await request("diary/\(id)", method: "PUT", body: payload) }
    func deleteDiary(id: Int) async throws -> DeleteResponse { try await request("diary/\(id)", method: "DELETE") }
    func createProgress(_ payload: ProgressPayload) async throws -> ProgressEntry { try await request("progress", method: "POST", body: payload) }
    func createProgress(_ payload: JSONPayload) async throws -> ProgressEntry { try await request("progress", method: "POST", body: payload) }
    func updateProgress(id: Int, payload: JSONPayload) async throws -> ProgressEntry { try await request("progress/\(id)", method: "PUT", body: payload) }
    func deleteProgress(id: Int) async throws -> DeleteResponse { try await request("progress/\(id)", method: "DELETE") }
    func createWorkout(_ payload: WorkoutPayload) async throws -> WorkoutEntry { try await request("workouts", method: "POST", body: payload) }
    func createWorkout(_ payload: JSONPayload) async throws -> WorkoutEntry { try await request("workouts", method: "POST", body: payload) }
    func completeWorkoutPlan(id: Int) async throws -> WorkoutPlan { try await request("workout-plans/\(id)/complete", method: "POST") }
    func logout() async {
        do { let _: EmptyResponse = try await request("auth/logout", method: "POST") }
        catch { /* JWT is invalidated locally below. */ }
    }
}

private struct AuthBody: Encodable { let email: String; let password: String }
struct DeleteResponse: Decodable { let deleted: Bool; let id: Int }
struct EmptyResponse: Decodable { let ok: Bool }
struct TrainerInfoResponse: Codable { let trainer: TrainerInfo? }

struct JSONPayload: Encodable {
    let values: [String: AnyEncodable]
    private struct CodingKeyImpl: CodingKey {
        let stringValue: String
        init?(stringValue: String) { self.stringValue = stringValue }
        let intValue: Int? = nil
        init?(intValue: Int) { return nil }
    }
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeyImpl.self)
        for (key, value) in values { try container.encode(value, forKey: CodingKeyImpl(stringValue: key)!) }
    }
}

struct AnyEncodable: Encodable {
    private let encodeClosure: (Encoder) throws -> Void
    init(_ value: Encodable) { encodeClosure = { encoder in try value.encode(to: encoder) } }
    func encode(to encoder: Encoder) throws { try encodeClosure(encoder) }
}

File: mobile/android/app/src/main/java/com/astra/nutrition/ApiClient.kt
package com.astra.nutrition

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class ApiClient(context: Context) {
    private val prefs = context.getSharedPreferences("astra", Context.MODE_PRIVATE)
    private val defaultApiUrl = "http://204.168.255.69:8787/api/v1"
    private val legacyLocalApiUrls = setOf("http://127.0.0.1:8787/api/v1", "http://10.0.2.2:8787/api/v1")
    var baseUrl: String
        get() {
            val stored = prefs.getString("api_url", null)
            return if (stored == null || stored in legacyLocalApiUrls) defaultApiUrl else stored
        }
        set(value) { prefs.edit().putString("api_url", value.trim().trimEnd('/')).apply() }
    var token: String?
        get() = prefs.getString("access_token", null)
        set(value) { prefs.edit().apply { if (value == null) remove("access_token") else putString("access_token", value) }.apply() }

    private suspend fun request(path: String, method: String = "GET", body: JSONObject? = null): Any = withContext(Dispatchers.IO) {
        val connection = (URL("${baseUrl.trimEnd('/')}/$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15_000
            readTimeout = 30_000
            setRequestProperty("Accept", "application/json")
            token?.let { setRequestProperty("Authorization", "Bearer $it") }
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                outputStream.use { it.write(body.toString().toByteArray()) }
            }
        }
        try {
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (status !in 200..299) {
                val message = runCatching { JSONObject(text).optString("error") }.getOrNull().orEmpty()
                throw AstraException(message.ifBlank { "Ошибка запроса ($status)" })
            }
            if (text.isBlank()) JSONObject() else if (text.trimStart().startsWith("[")) JSONArray(text) else JSONObject(text)
        } finally { connection.disconnect() }
    }

    suspend fun login(email: String, password: String, register: Boolean = false): AuthResponse {
        val result = request("auth/${if (register) "register" else "login"}", "POST", JSONObject().put("email", email).put("password", password)) as JSONObject
        return AuthResponse(result.optString("access_token"), result.optJSONObject("user")!!.user())
    }
    suspend fun me(): AuthUser = (request("auth/me") as JSONObject).user()
    suspend fun logout() {
        try { request("auth/logout", "POST") } catch (_: Throwable) { /* JWT is cleared locally below. */ }
        token = null
    }
    suspend fun dashboard(): Dashboard { val o = request("dashboard") as JSONObject; return Dashboard(o.optInt("products"), o.optInt("recipes"), o.optInt("approved"), o.optJSONObject("latest")?.progress(), o.optJSONArray("top").recipes()) }
    suspend fun products(): List<Product> = (request("products") as JSONArray).products()
    suspend fun recipes(): List<Recipe> = (request("recipes") as JSONArray).recipes()
    suspend fun recipe(id: Int): RecipeDetail { val o = request("recipes/$id") as JSONObject; return RecipeDetail(o.optJSONObject("recipe")!!.recipe(), o.optJSONArray("ingredients").ingredients()) }
    suspend fun diary(): List<DiaryEntry> = (request("diary") as JSONArray).diary()
    suspend fun progress(): List<ProgressEntry> = (request("progress") as JSONArray).progresses()
    suspend fun workouts(): List<WorkoutEntry> = (request("workouts") as JSONArray).workouts()
    suspend fun plans(): List<WorkoutPlan> = (request("workout-plans") as JSONArray).plans()
    suspend fun exercises(): List<Exercise> = (request("exercises") as JSONArray).exercises()
    suspend fun workoutEquipment(): List<WorkoutEquipment> = (request("workout-equipment") as JSONArray).equipment()
    suspend fun workoutComplexes(): List<WorkoutComplex> = (request("workout-complexes") as JSONArray).complexes()
    suspend fun categories(kind: String): List<ContentCategory> = (request("categories?kind=$kind") as JSONArray).categories()
    suspend fun articleSections(): List<ArticleSection> = (request("article-sections") as JSONArray).articleSections()
    suspend fun articles(): List<Article> = (request("articles") as JSONArray).articles()
    suspend fun myTrainerChat(): TrainerChatResponse { val o = request("clients/me/chat") as JSONObject; return o.trainerInfo() }
    suspend fun sendMyTrainerChat(message: String): TrainerChatMessage = (request("clients/me/chat", "POST", JSONObject().put("message", message)) as JSONObject).chatMessage()
    suspend fun clients(): List<ClientSummary> = (request("clients") as JSONArray).clients()
    suspend fun client(id: Int): ClientDetail = (request("clients/$id") as JSONObject).clientDetail()
    suspend fun clientChat(id: Int): List<TrainerChatMessage> = (request("clients/$id/chat") as JSONArray).chatMessages()
    suspend fun sendClientChat(id: Int, message: String): TrainerChatMessage = (request("clients/$id/chat", "POST", JSONObject().put("message", message)) as JSONObject).chatMessage()

    suspend fun createProduct(body: JSONObject): Product = (request("products", "POST", body) as JSONObject).product()
    suspend fun updateProduct(id: Int, body: JSONObject): Product = (request("products/$id", "PUT", body) as JSONObject).product()
    suspend fun deleteProduct(id: Int) { request("products/$id", "DELETE") }
    suspend fun createRecipe(body: JSONObject): Recipe = (request("recipes", "POST", body) as JSONObject).recipe()
    suspend fun updateRecipe(id: Int, body: JSONObject): Recipe = (request("recipes/$id", "PUT", body) as JSONObject).recipe()
    suspend fun deleteRecipe(id: Int) { request("recipes/$id", "DELETE") }
    suspend fun requestRecipeSubmission(id: Int): Recipe = (request("recipes/$id/submission-request", "POST") as JSONObject).recipe()
    suspend fun cancelRecipeSubmission(id: Int): Recipe = (request("recipes/$id/submission-request", "DELETE") as JSONObject).recipe()
    suspend fun moderateRecipe(id: Int, action: String, note: String?): Recipe = (request("recipes/$id/moderation", "POST", JSONObject().put("action", action).putNullable("note", note)) as JSONObject).recipe()
    suspend fun createExercise(body: JSONObject): Exercise = (request("exercises", "POST", body) as JSONObject).exercise()
    suspend fun updateExercise(id: Int, body: JSONObject): Exercise = (request("exercises/$id", "PUT", body) as JSONObject).exercise()
    suspend fun deleteExercise(id: Int) { request("exercises/$id", "DELETE") }
    suspend fun createEquipment(body: JSONObject): WorkoutEquipment = (request("workout-equipment", "POST", body) as JSONObject).equipment()
    suspend fun updateEquipment(id: Int, body: JSONObject): WorkoutEquipment = (request("workout-equipment/$id", "PUT", body) as JSONObject).equipment()
    suspend fun deleteEquipment(id: Int) { request("workout-equipment/$id", "DELETE") }
    suspend fun createComplex(body: JSONObject): WorkoutComplex = (request("workout-complexes", "POST", body) as JSONObject).complex()
    suspend fun updateComplex(id: Int, body: JSONObject): WorkoutComplex = (request("workout-complexes/$id", "PUT", body) as JSONObject).complex()
    suspend fun updateWorkout(id: Int, body: JSONObject): WorkoutEntry = (request("workouts/$id", "PUT", body) as JSONObject).workout()
    suspend fun deleteWorkout(id: Int) { request("workouts/$id", "DELETE") }
    suspend fun createPlan(body: JSONObject): WorkoutPlan = (request("workout-plans", "POST", body) as JSONObject).plan()
    suspend fun updatePlan(id: Int, body: JSONObject): WorkoutPlan = (request("workout-plans/$id", "PUT", body) as JSONObject).plan()
    suspend fun cancelPlan(id: Int): WorkoutPlan = (request("workout-plans/$id/cancel", "POST") as JSONObject).plan()
    suspend fun deletePlan(id: Int) { request("workout-plans/$id", "DELETE") }
    suspend fun createArticle(body: JSONObject): Article = (request("articles", "POST", body) as JSONObject).article()
    suspend fun updateArticle(id: Int, body: JSONObject): Article = (request("articles/$id", "PUT", body) as JSONObject).article()
    suspend fun updateArticleFlags(id: Int, body: JSONObject): Article = (request("articles/$id/flags", "PATCH", body) as JSONObject).article()
    suspend fun deleteArticle(id: Int) { request("articles/$id", "DELETE") }
    suspend fun shareToTrainer(type: String, id: Int): JSONObject = request("clients/me/shares", "POST", JSONObject().put("item_type", type).put("item_id", id)) as JSONObject
    suspend fun shareToClient(clientId: Int, type: String, id: Int): JSONObject = request("clients/shares", "POST", JSONObject().put("client_id", clientId).put("item_type", type).put("item_id", id)) as JSONObject

    suspend fun addDiary(date: String, meal: String, productId: Int?, recipeId: Int?, amount: Double): List<DiaryEntry> {
        val item = JSONObject().put("meal_type", meal).put("servings", if (recipeId != null) amount else 1.0)
        if (productId != null) item.put("product_id", productId).put("quantity", amount) else item.put("recipe_id", recipeId)
        return (request("diary", "POST", JSONObject().put("entry_date", date).put("items", JSONArray().put(item))) as JSONArray).diary()
    }
    suspend fun addDiary(body: JSONObject): List<DiaryEntry> = (request("diary", "POST", body) as JSONArray).diary()
    suspend fun updateDiary(id: Int, body: JSONObject): DiaryEntry = (request("diary/$id", "PUT", body) as JSONObject).diary()
    suspend fun deleteDiary(id: Int) { request("diary/$id", "DELETE") }
    suspend fun addProgress(date: String, weight: Double?, waist: Double?, wellbeing: Double?, comment: String?): ProgressEntry {
        val body = JSONObject().put("measured_at", date).putNullable("weight_kg", weight).putNullable("waist_cm", waist).putNullable("wellbeing_score", wellbeing).putNullable("comment", comment)
        return (request("progress", "POST", body) as JSONObject).progress()
    }
    suspend fun createProgress(body: JSONObject): ProgressEntry = (request("progress", "POST", body) as JSONObject).progress()
    suspend fun updateProgress(id: Int, body: JSONObject): ProgressEntry = (request("progress/$id", "PUT", body) as JSONObject).progress()
    suspend fun deleteProgress(id: Int) { request("progress/$id", "DELETE") }
    suspend fun addWorkout(date: String, exerciseId: Int, weight: Double?, sets: Double?, reps: Double?, rir: String?): WorkoutEntry {
        val body = JSONObject().put("performed_at", date).put("exercise_id", exerciseId).putNullable("working_weight", weight).putNullable("sets", sets).putNullable("reps", reps).putNullable("rir", rir)
        return (request("workouts", "POST", body) as JSONObject).workout()
    }
    suspend fun createWorkout(body: JSONObject): WorkoutEntry = (request("workouts", "POST", body) as JSONObject).workout()
    suspend fun completePlan(id: Int): WorkoutPlan = (request("workout-plans/$id/complete", "POST") as JSONObject).plan()
}

private fun JSONObject.putNullable(key: String, value: Any?): JSONObject = put(key, value ?: JSONObject.NULL)
private fun JSONObject.optDoubleOrNull(key: String): Double? = if (!has(key) || isNull(key)) null else optDouble(key)
private fun JSONObject.optTextOrNull(key: String): String? = optString(key).takeIf { it.isNotBlank() }
private fun JSONObject.user() = AuthUser(optInt("id"), optString("email"), optString("name"), optBoolean("is_admin"), optBoolean("is_trainer"))
private fun JSONObject.recipe() = Recipe(optInt("id"), optString("code"), optString("name"), optString("category"), optTextOrNull("status"), optDoubleOrNull("servings"), optDoubleOrNull("kcal_per_serving"), optDoubleOrNull("protein_per_serving_g"), optDoubleOrNull("fat_per_serving_g"), optDoubleOrNull("carbs_per_serving_g"), optDoubleOrNull("cost_per_serving_rsd"), optTextOrNull("subcategory"), optTextOrNull("version"), optTextOrNull("tags"), optBoolean("is_ready"), optBoolean("needs_garnish"), optTextOrNull("collection"), optInt("owner_id").takeIf { it > 0 }, optBoolean("submission_requested"), optTextOrNull("moderation_status"), optTextOrNull("moderation_note"), optBoolean("is_submitter"))
private fun JSONObject.product() = Product(optInt("id"), optString("code"), optString("name"), optTextOrNull("category"), optTextOrNull("unit"), optDoubleOrNull("kcal"), optDoubleOrNull("protein_g"), optDoubleOrNull("fat_g"), optDoubleOrNull("carbs_g"), optDoubleOrNull("package_price_rsd"), optDoubleOrNull("package_size"), optDoubleOrNull("price_per_100_or_unit_rsd"), optTextOrNull("data_status"), optTextOrNull("note"))
private fun JSONObject.ingredient() = Ingredient(optInt("id"), optInt("product_id"), optString("name"), optDoubleOrNull("quantity"), optTextOrNull("unit"), optTextOrNull("portion_description"), optDoubleOrNull("kcal"), optDoubleOrNull("protein_g"), optDoubleOrNull("fat_g"), optDoubleOrNull("carbs_g"), optDoubleOrNull("cost_rsd"))
private fun JSONObject.diary() = DiaryEntry(optInt("id"), optString("entry_date"), optTextOrNull("meal_type"), optInt("recipe_id").takeIf { it > 0 }, optInt("product_id").takeIf { it > 0 }, optDoubleOrNull("servings"), optDoubleOrNull("quantity"), optTextOrNull("unit"), optTextOrNull("measurement_name"), optDoubleOrNull("measurement_quantity"), optTextOrNull("comment"), optTextOrNull("name"), optString("item_type"), optDoubleOrNull("kcal_per_serving"), optDoubleOrNull("protein_per_serving_g"), optDoubleOrNull("fat_per_serving_g"), optDoubleOrNull("carbs_per_serving_g"), optDoubleOrNull("cost_per_serving_rsd"))
private fun JSONObject.progress() = ProgressEntry(optInt("id"), optString("measured_at"), optDoubleOrNull("weight_kg"), optDoubleOrNull("desired_weight_kg"), optDoubleOrNull("height_cm"), optDoubleOrNull("bmi"), optDoubleOrNull("body_fat_pct"), optDoubleOrNull("muscle_mass_kg"), optDoubleOrNull("kcal_target"), optDoubleOrNull("protein_target_g"), optDoubleOrNull("fat_target_g"), optDoubleOrNull("carbs_target_g"), optDoubleOrNull("waist_cm"), optDoubleOrNull("chest_cm"), optDoubleOrNull("hips_cm"), optDoubleOrNull("sleep_score"), optDoubleOrNull("wellbeing_score"), optTextOrNull("comment"))
private fun JSONObject.planItem() = WorkoutPlanItem(if (has("id") && !isNull("id")) optInt("id") else null, optInt("exercise_id"), optTextOrNull("name"), optTextOrNull("muscle_group"), optDoubleOrNull("working_weight"), optDoubleOrNull("sets"), optDoubleOrNull("duration_minutes"), optDoubleOrNull("speed_kmh"))
private fun JSONObject.plan() = WorkoutPlan(optInt("id"), optString("scheduled_at"), optDoubleOrNull("duration_minutes"), optString("status"), optTextOrNull("completed_at"), optJSONArray("items").planItems())
private fun JSONObject.workout() = WorkoutEntry(optInt("id"), optString("performed_at"), optInt("exercise_id"), optString("name"), optTextOrNull("muscle_group"), optDoubleOrNull("working_weight"), optDoubleOrNull("sets"), optDoubleOrNull("reps"), optTextOrNull("rir"), optTextOrNull("comment"))
private fun JSONObject.exerciseVariant() = ExerciseVariant(optTextOrNull("machine"), optTextOrNull("equipment"), optTextOrNull("description"), optTextOrNull("technique"), optTextOrNull("tips"))
private fun JSONObject.exercise() = Exercise(optInt("id"), optString("name"), optTextOrNull("muscle_group"), optTextOrNull("default_unit"), optDoubleOrNull("default_sets"), optDoubleOrNull("default_reps"), optTextOrNull("target_rir"), optTextOrNull("note"), optTextOrNull("description"), optJSONArray("photos").strings(), optTextOrNull("video"), optJSONArray("variants").exerciseVariants())
private fun JSONObject.equipment() = WorkoutEquipment(optInt("id"), optString("kind"), optString("name"), optTextOrNull("description"), optTextOrNull("photo"))
private fun JSONObject.complexItem() = WorkoutComplexItem(optInt("id"), optInt("exercise_id"), optTextOrNull("exercise_code"), optString("name"), optTextOrNull("muscle_group"), optTextOrNull("default_unit"), optDoubleOrNull("working_weight"), optDoubleOrNull("sets"), optDoubleOrNull("duration_minutes"), optDoubleOrNull("speed_kmh"))
private fun JSONObject.complex() = WorkoutComplex(optInt("id"), optString("name"), optTextOrNull("comment"), optJSONArray("photos").strings(), optTextOrNull("video"), optJSONArray("items").complexItems())
private fun JSONObject.category() = ContentCategory(optInt("id"), optString("kind"), optString("name"), optString("collection"), optInt("owner_id").takeIf { it > 0 })
private fun JSONObject.articleSection() = ArticleSection(optInt("id"), optString("name"), optTextOrNull("description"), optInt("article_count"))
private fun JSONObject.articleLink() = ArticleLink(optString("title"), optString("url"))
private fun JSONObject.article() = Article(optInt("id"), optInt("section_id"), optString("section_name"), optString("title"), optString("body"), optTextOrNull("tags"), optJSONArray("links").articleLinks(), optJSONArray("photos").strings(), optTextOrNull("video"), optBoolean("is_pinned"), optBoolean("is_hidden"), optString("created_at"), optTextOrNull("updated_at"))
private fun JSONObject.trainerInfo() = TrainerChatResponse(optJSONObject("trainer")?.trainer(), optJSONArray("messages").chatMessages(), optInt("unread_count"))
private fun JSONObject.trainer() = TrainerInfo(optInt("id"), optString("name"), optString("email"))
private fun JSONObject.sharedItem() = SharedChatItem(optString("type"), optInt("id"), optString("name"))
private fun JSONObject.chatMessage() = TrainerChatMessage(optInt("id"), optInt("sender_id").takeIf { it > 0 }, optString("sender_name"), optString("message"), optJSONObject("shared_item")?.sharedItem(), optString("created_at"))
private fun JSONObject.clientNextWorkout() = ClientNextWorkout(optInt("id"), optString("scheduled_at"), optString("status"))
private fun JSONObject.clientSummary() = ClientSummary(optInt("id"), optString("name"), optString("email"), optJSONObject("next_workout")?.clientNextWorkout(), optInt("unread_messages"))
private fun JSONObject.clientNutrition() = ClientNutritionValues(optDoubleOrNull("kcal"), optDoubleOrNull("protein"), optDoubleOrNull("fat"), optDoubleOrNull("carbs"))
private fun JSONObject.clientToday() = ClientToday(optString("date"), optJSONArray("entries").diary(), optJSONObject("totals")?.clientNutrition() ?: ClientNutritionValues(0.0, 0.0, 0.0, 0.0), optJSONObject("targets")?.clientNutrition() ?: ClientNutritionValues(null, null, null, null), optJSONObject("remaining")?.clientNutrition() ?: ClientNutritionValues(null, null, null, null))
private fun JSONObject.clientDetail() = ClientDetail(optInt("id"), optString("name"), optString("email"), optJSONObject("next_workout")?.clientNextWorkout(), optInt("unread_messages"), optJSONArray("progress").progresses(), optJSONObject("today")?.clientToday() ?: ClientToday("", emptyList(), ClientNutritionValues(0.0, 0.0, 0.0, 0.0), ClientNutritionValues(null, null, null, null), ClientNutritionValues(null, null, null, null)), optJSONArray("workouts").workouts(), optJSONArray("workout_plans").plans())
private fun JSONArray?.recipes() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.recipe() }
private fun JSONArray?.products() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.product() }
private fun JSONArray?.ingredients() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.ingredient() }
private fun JSONArray?.diary() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.diary() }
private fun JSONArray?.progresses() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.progress() }
private fun JSONArray?.plans() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.plan() }
private fun JSONArray?.planItems() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.planItem() }
private fun JSONArray?.workouts() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.workout() }
private fun JSONArray?.exercises() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.exercise() }
private fun JSONArray?.equipment() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.equipment() }
private fun JSONArray?.complexItems() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.complexItem() }
private fun JSONArray?.exerciseVariants() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.exerciseVariant() }
private fun JSONArray?.complexes() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.complex() }
private fun JSONArray?.categories() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.category() }
private fun JSONArray?.articleSections() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.articleSection() }
private fun JSONArray?.articleLinks() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.articleLink() }
private fun JSONArray?.chatMessages() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.chatMessage() }
private fun JSONArray?.clients() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.clientSummary() }
private fun JSONArray?.strings() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optString(it)?.takeIf { value -> value.isNotBlank() } }
private fun JSONArray?.articles() = (0 until (this?.length() ?: 0)).mapNotNull { this?.optJSONObject(it)?.article() }
